File: backend/spec/factories/notifications.rb
FactoryBot.define do
  factory :notification do
    user
    title { Faker::Lorem.sentence(word_count: 3) }
    message { Faker::Lorem.paragraph(sentence_count: 2) }
    notification_type { 'info' }
    delivery_channels { ['in_app'] }
    event_type { 'test_event' }
    event_id { nil }
    metadata { {} }
    delivered { false }
    delivered_at { nil }
    read_at { nil }

    trait :read do
      read_at { Time.current }
    end

    trait :unread do
      read_at { nil }
    end

    trait :delivered do
      delivered { true }
      delivered_at { Time.current }
    end

    trait :pending_delivery do
      delivered { false }
      delivered_at { nil }
    end

    trait :with_email do
      delivery_channels { ['in_app', 'email'] }
    end

    trait :admin_message do
      notification_type { 'admin_message' }
      event_type { 'admin_message' }
    end

    trait :booking_update do
      notification_type { 'booking_update' }
      event_type { 'booking_created' }
      event_id { SecureRandom.hex(8) }
    end

    trait :system do
      notification_type { 'system' }
      event_type { 'system_maintenance' }
    end

    trait :with_metadata do
      metadata { { 
        source: 'test',
        priority: 'high',
        category: 'user_action'
      } }
    end
  end
end

File: frontend/src/types/notifications.ts
// Types for notification system

export interface Notification {
  id: number
  title: string
  message: string
  type: NotificationType
  delivery_channels: DeliveryChannel[]
  read: boolean
  read_at: string | null
  created_at: string
  metadata: Record<string, any>
}

export interface AdminNotification extends Notification {
  user: {
    id: number
    email: string
    name: string
  }
  event_type: string
  event_id: string | null
  delivered: boolean
  delivered_at: string | null
}

export type NotificationType = 
  | 'info' 
  | 'success' 
  | 'warning' 
  | 'error' 
  | 'booking_update' 
  | 'system' 
  | 'admin_message'

export type DeliveryChannel = 
  | 'in_app' 
  | 'email' 
  | 'sms' 
  | 'push' 
  | 'webhook'

export interface NotificationFormData {
  title: string
  message: string
  notification_type: NotificationType
  delivery_channels: DeliveryChannel[]
  metadata?: Record<string, any>
}

export interface BulkNotificationData {
  notification: NotificationFormData
  user_ids: number[]
  delivery_channels: DeliveryChannel[]
}

export interface NotificationStats {
  total: number
  unread: number
  delivered: number
  pending_delivery: number
  by_type: Record<NotificationType, number>
  by_event: Record<string, number>
  recent_24h: number
}

export interface DeliveryStats {
  total_notifications: number
  delivered_notifications: number
  pending_notifications: number
  channel_stats: Record<DeliveryChannel, {
    total: number
    delivered: number
    pending: number
  }>
  error_rate: number
}

export interface NotificationFilters {
  type?: NotificationType
  event_type?: string
  user_id?: number
  delivered?: boolean
  page?: number
  per_page?: number
}

export interface NotificationResponse {
  success: boolean
  notifications: Notification[]
  pagination: {
    current_page: number
    total_pages: number
    total_count: number
    per_page: number
  }
  unread_count: number
}

export interface AdminNotificationResponse {
  success: boolean
  notifications: AdminNotification[]
  pagination: {
    current_page: number
    total_pages: number
    total_count: number
    per_page: number
  }
  stats: NotificationStats
}

export interface BulkNotificationResult {
  user_id: number
  success: boolean
  notification_id?: number
  error?: string
}

export interface BulkNotificationResponse {
  success: boolean
  message: string
  results: BulkNotificationResult[]
  stats: {
    total: number
    successful: number
    failed: number
  }
}
